functionary/train/train_lora.py

```diff
@@ -317,6 +327,41 @@ def initialize_tokenizer(
     return tokenizer
 
 
+def read_dataset(data_args: DataArguments, training_args: TrainingArguments, tokenizer: Any, ds_type: str):
+    ds_class = CustomDataset
+    if data_args.packing:
+        ds_class = PackedDataset
+        
+    cached_folder = os.path.join(training_args.output_dir, f"{ds_type}_cached")
+    
+    if training_args.local_rank > 0:
+        print(f"process: {LOCAL_RANK} wait for main process to prepare the training data")
+        torch.distributed.barrier()
+    else:  # rank 0 will process data first, others will wait
+        if not os.path.exists(training_args.output_dir):
+            os.mkdir(training_args.output_dir)
+        if not os.path.exists(cached_folder):
+            os.mkdir(cached_folder)
+            
+        data_path = data_args.train_data_path if ds_type == "train" else data_args.eval_data_path
+        data_ratio = data_args.training_ratio if ds_type == "train" else data_args.eval_ratio
+        with open(data_path, "r") as file:
+            raw_train_data = [json.loads(line) for line in file]
+            if data_ratio < 1:
+                raw_train_data = raw_train_data[: int(data_ratio * len(raw_train_data))]
+        print(f"{ds_type} size: : {len(raw_train_data)}")
+        # set ignore_cached=True to process data at rank 0
+        ds = ds_class(raw_train_data, tokenizer, cached_folder=cached_folder, ignore_cached=True)
+        print(f"process: {LOCAL_RANK} finish processing data")
+        torch.distributed.barrier()
+        
+    # All ranks will read the processed data from cached_path created by rank 0
+    ds = ds_class(None, tokenizer, cached_folder=cached_folder, ignore_cached=False)
+    if LOCAL_RANK == 0:
+        ds.stat()
+    return ds
```

Comment:
Can we add more spacing and concise comment in between the sections for this function, for better readability

Reply:
Sure